disnake/enums.py

```diff
@@ -72,326 +85,1370 @@
-        _enum_value_map_: ClassVar[Dict[Any, Any]]
-
-    def __new__(cls, name, bases, attrs, *, comparable: bool = False):
-        value_mapping = {}
-        member_mapping = {}
-        member_names = []
-
-        value_cls = _create_value_cls(name, comparable)
-        for key, value in list(attrs.items()):
-            is_descriptor = _is_descriptor(value)
-            if key[0] == "_" and not is_descriptor:
-                continue
-
-            # Special case classmethod to just pass through
-            if isinstance(value, classmethod):
-                continue
-
-            if is_descriptor:
-                setattr(value_cls, key, value)
-                del attrs[key]
-                continue
-
-            try:
-                new_value = value_mapping[value]
-            except KeyError:
-                new_value = value_cls(name=key, value=value)
-                value_mapping[value] = new_value
-                member_names.append(key)
-
-            member_mapping[key] = new_value
-            attrs[key] = new_value
-
-        attrs["_enum_value_map_"] = value_mapping
-        attrs["_enum_member_map_"] = member_mapping
-        attrs["_enum_member_names_"] = member_names
-        attrs["_enum_value_cls_"] = value_cls
-        actual_cls = super().__new__(cls, name, bases, attrs)
-        value_cls._actual_enum_cls_ = actual_cls  # type: ignore
-        return actual_cls
-
-    def __iter__(cls):
-        return (cls._enum_member_map_[name] for name in cls._enum_member_names_)
-
-    def __reversed__(cls):
-        return (cls._enum_member_map_[name] for name in reversed(cls._enum_member_names_))
-
-    def __len__(cls):
-        return len(cls._enum_member_names_)
-
-    def __repr__(cls):
+    __is_enum_instantiated: ClassVar[bool] = False
+
+    _name_map_: ClassVar[Mapping[str, Enum]]
+    _value2member_map_: ClassVar[Mapping[Any, Enum]]
+
+    def __new__(
+        metacls: Type[_T],  # pyright: reportSelfClsParameterName=false
+        name: str,
+        bases: Tuple[Type[Any], Type[Any]],
+        namespace: _EnumDict,
+    ) -> _T:
+
+        if not EnumMeta.__is_enum_instantiated:
+            EnumMeta.__is_enum_instantiated = True
+            return super().__new__(metacls, name, bases, namespace)
+
+        base, enum_type = bases  # ensured possible in __prepare__
+
+        ns: Dict[str, Any] = {
+            "__objtype__": base,
+            "__enumtype__": enum_type,
+            "_name_map_": (name_map := {}),
+            "_value2member_map_": (value_map := {}),
+            **{
+                name_: value_
+                for name_, value_ in Enum.__dict__.items()
+                if name_ not in ("__class__", "__module__", "__doc__")
+            },
+        }
+
+        ns.update(namespace)
+
+        cls = super().__new__(metacls, name, bases, ns)
+
+        for name_, value_ in namespace.member_map.items():
+            member = cls.__new__(cls, value_)  # type: ignore
```

Comment:
The previous implementation reused member objects with identical values, is there a reason for doing one or the other?
One disadvantage of reusing objects is that `SomeEnum.abc.name` isn't necessarily `"abc"`, but I can't think of anything else :thinking:

Reply:
This too was mainly done to mirror the default implementation. We could definitely decide to reuse duplicates to save memory, though. It ultimately wouldn't matter too much for our use case.

Reply:
After some discussion on discord, the decision was made to reuse existing member objects as suggested. This was implemented as part of f8c0c07f1f2adf5cc34607d90f1b6718a694ec05.